### PR TITLE
i#4014 dr$sim phys: Make physaddr_t per-thread

### PR DESCRIPTION
**Pull request by derekbruening**
The physaddr_t class is not thread-safe and was previously used racily
in the drmemtrace code.  We fix that by creating a separate instance
per thread.  A test with multiple threads is added.

Issue: #4014

### CODE REVIEW

File: clients/drcachesim/tracer/tracer.cpp [hunk 1]
```diff
@@ -2662,7 +2671,7 @@ event_thread_init(void *drcontext)
 {
     per_thread_t *data = (per_thread_t *)dr_thread_alloc(drcontext, sizeof(per_thread_t));
     DR_ASSERT(data != NULL);
-    memset(data, 0, sizeof(*data));
+    *data = {};
```
Comment (abhinav92003) [inline clients/drcachesim/tracer/tracer.cpp:2674]: Why was this change needed?

File: clients/drcachesim/tracer/physaddr.h [hunk 2]
```diff
@@ -52,10 +54,11 @@ class physaddr_t {
     virtual2physical(addr_t virt);
 
 private:
-    // Assumed to be single-threaded
 #ifdef LINUX
     addr_t last_vpage_;
     addr_t last_ppage_;
+    // XXX: An app with thousands of threads might hit open file limits.
```
Comment (abhinav92003) [inline clients/drcachesim/tracer/physaddr.h:60]: Does this use one of DR's private FDs? I don't think so, as the implementation doesn't use `dr_file_open`. It probably should right? But then I'd be concerned about exhausting the small number of available private FDs (96). A lock would cause too much contention I suppose, so that's not the solution...

File: clients/drcachesim/tracer/tracer.cpp (hunk 1 above)
Comment (derekbruening) [reply clients/drcachesim/tracer/tracer.cpp:2674]: The compiler warns about zeroing a no-longer-POD struct: this is how you're supposed to zero w/o zeroing the class (and potentially clobbering vtable).

File: clients/drcachesim/tracer/physaddr.h (hunk 2 above)
Comment (derekbruening) [reply clients/drcachesim/tracer/physaddr.h:60]: It does use a private FD: "open" is redirected to "os_open_protected".  So a many-threaded app will indeed use that up.  Hmm.

Comment (abhinav92003) [reply clients/drcachesim/tracer/physaddr.h:60]: > "open" is redirected to "os_open_protected"

I thought only the app's `open`s are redirected.

Comment (derekbruening) [reply clients/drcachesim/tracer/physaddr.h:60]: "Redirected" means pointing to a different function, done by the private loader to client and private libraries: https://github.com/DynamoRIO/dynamorio/blob/master/core/unix/loader.c#L1438

For the app using SYS_open: I don't think DR does anything there.  Looking: just some debug stuff and some emulation stuff.

Comment (abhinav92003) [reply clients/drcachesim/tracer/physaddr.h:60]: > For the app using SYS_open

Oh right, got it.

> "Redirected" means pointing to a different function, done by the private loader to client and private libraries:

Okay.

Comment (derekbruening) [reply clients/drcachesim/tracer/physaddr.h:60]: I think for now I will augment the comment to talk about using up the private FDs but not try to solve it yet; the goal here is to get simple apps/benchmarks working.  We'll list scaling up the thousands of threads as a separate issue to solve.

Comment (derekbruening) [reply clients/drcachesim/tracer/physaddr.h]: Expanded the comment pointing to #4014.

File: clients/drcachesim/tracer/tracer.cpp (hunk 1 above)
Comment (derekbruening) [reply clients/drcachesim/tracer/tracer.cpp]: Added a comment.

**Review by abhinav92003 — APPROVED** [review]
